include/frg/bitset.hpp

```diff
@@ -0,0 +1,263 @@
+                size_t offset = pos % 64;
+                if (offset == 0)
+                    for (size_t i = N - 1; i >= wshift; --i)
+                        buffer[i] = buffer[i - wshift];
+                else
+                {
+                    const size_t soffset = (64 - offset);
+                    for (size_t i = N - 1; i > wshift; --i)
+                        buffer[i] = ((buffer[i - wshift] << offset) | (buffer[i - wshift - 1] >> soffset));
+                    buffer[wshift] = buffer[0] << offset;
+                }
+
+                for (auto i = buffer + 0; i < buffer + wshift; i++)
+                    *i = 0;
+            }
+
+            return *this;
+        }
+
+        bitset& operator>>=(size_t pos) noexcept
+        {
+            if (pos != 0)
+            {
+                const size_t wshift = pos / 64;
+                const size_t offset = pos % 64;
+                const size_t s = N - wshift - 1;
+
+                if (offset == 0)
+                    for (size_t i = 0; i <= s; ++i)
+                        buffer[i] = buffer[i + wshift];
+                else
+                {
+                    const size_t off = (64 - offset);
+                    for (size_t i = 0; i < s; ++i)
+                        buffer[i] = ((buffer[i + wshift] >> offset) | (buffer[i + wshift + 1] << off));
+                    buffer[s] = buffer[N - 1] >> offset;
+                }
+
+                for (auto i = buffer + s + 1; i < buffer + N; i++)
+                    *i = 0;
+            }
+
+            return *this;
+        }
+
+        bitset& set() noexcept
+        {
+            for (auto& i : buffer)
+                i = ~0;
+            return *this;
+        }
+
+        bitset& set(size_t pos, bool val = true)
+        {
+            set_bit(buffer[pos / 64], val, pos % 64);
+            return *this;
+        }
+
+        bitset& reset() noexcept
+        {
+            for (auto& i : buffer)
+                i = 0;
+            return *this;
+        }
+
+        bitset& reset(size_t pos) { return set(pos, false); }
+
+        bitset operator~() const noexcept
+        {
+            auto copy = *this;
+            copy.flip();
+            return copy;
+        }
+
+        bitset& flip() noexcept
+        {
+            for (auto& i : buffer)
+                i = ~i;
+            return *this;
+        }
+
+        bitset& flip(size_t pos) { return set(pos, !this->operator[](pos)); }
+
+        // element access
+        constexpr bool operator[](size_t pos) const { return get_bit(buffer[pos / 64], pos % 64); }
+
+        reference operator[](size_t pos) { return reference(pos, *this); }
+
+        size_t count() const noexcept
+        {
+            size_t n = 0;
+            for (size_t i = 0; i < buffer_size - 1; i++)
+                n += __builtin_popcountll(buffer[i]);
+            return n + __builtin_popcountll(get_bits(buffer[buffer_size - 1], 0, N % 64));
```

Comment:
This probably isn't needed (the upper bits of this bufer element should always be zero)